TESTS/mbed_platform/FileHandle/main.cpp

```diff
@@ -0,0 +1,425 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "greentea-client/test_env.h"
+#include "utest/utest.h"
+#include "unity/unity.h"
+#include "TestFile.h"
+#include "mbed.h"
+
+using utest::v1::Case;
+
+//#define PARTIAL_WRITE
```

Comment:
So, what would it take to get this on? Do we need any tweaks from #5571? IIRC, `_write()` for ARMCC is currently broken because it returns `size - written`, even when `written` is actually an error code.

Maybe we might want to split this into testing EOF for read, and actual errors on read/write. Those are key things that are easy to screw up, and critically important for a minority of users. So I think need to be in this test.

(Seen that in networking - very often TCP glue code can't do end-of-stream properly, which is fine for many users, but utterly breaks certain protocols).

Reply:
Firstly, maybe I should start testing with changes from #5571? 
Is there any sens of testing old (buggy) version while there is new one around the corner? 
When it's expected to be ready for merge?

Reply:
As far as I recall there was only that 1 intended adjustment to _write, but maybe you could just try merging with #5571 and cross-checking results.

At the present it's not clear whether there's a test error, retarget error or C library error.

I think you do need to lock the test down to be specifically testing EOF (read) and error (read or write), as those are the only "partial" returns I would expect the C library to handle. I haven't quite fully gone through to understand what the partial code is testing at present.

Other partial returns like non-blocking or special devices are needed, but they would require someone to be accessing the FileHandle directly (or using POSIX `read`/`write` from #5571) - I wouldn't expect to be able to do that through the C library stdio.

Reply:
If you have ifdeffed out bits that are failing - I'd like us to understand why, and I'd especially want to know if #5571 still doesn't fix them.

Ideally I'd like this to go in with just a few ifdeffed things that we know #5571 will fix (adjusting #5571 if necessary).

Reply:
Sure, the ifdefs are temporarily. 
Regarding write/read problem we have two scenarios:
1)	write more than file can handle - error occurs, we expects:
a.	`fwrite` should return number of objects written successfully
b.	`fwrite` set error state (checked then with `ferror`) when not all objects were written
2)	read more than file contain - `EOF` occurs
a.	`fread` should return number of objects read successfully
b.	`fread` set `EOF` when not all objects were read

I did some test on top of #5571 and get following results:
-	#5571 only fixes a bit IAR write behaviour (eliminates infinite write loop when write more than file can handle)
-	Described scenarios are fulfilled only on GCC_ARM compiler
-	ARM compiler results
o	scenario (1) :x: `fwrite` sets error, and return 0 instead of number of objects written successfully 
o	scenario (2) :white_check_mark: `fread` sets `EOF` and return number of objects read successfully 
-	IAR compiler results
o	scenario (1) :x: `fwrite` sets error, and return 3 instead of number of objects written successfully (5 to write, written 2, returned 3)
o	scenario (2) :white_check_mark: `fread` sets eof and return number of objects read successfully 

To sum up, looks like both ARM and IAR `fwrite` implementation is broken

Reply:
Thanks for the detailed work on this.

There is a third potential scenario - read gets an error.

Could you log the call parameters and return value from our retarget `_write` in scenario 1 for each toolchain to help me understand what's happening?

It's possible we may have to settle just for "error flag being set and number of objects returned less than requested" as the test - which is probably sufficient and what apps test for.

What about other calls like fputc - do they seem to work?

Reply:
Below log with write scenario
[log.txt](https://github.com/ARMmbed/mbed-os/files/1672765/log.txt)

fputc seems to be fine

Reply:
Thanks, you're a star! Will study and ponder.

Reply:
Okay, that does appear to be straightforward C library bugs. Don't think we could successfully work around those.

For ARM C, we are clearly returning the number of characters not written, as the documentation requires: "A positive number representing the number of characters not written (so any nonzero return value denotes a failure of some sort)."

The ARM C library detects this is an error as expected, but somehow doesn't pass back the correct return value. And because it doesn't make a final re-call like GCC, we won't get errno set either. But setting errno isn't strictly required by fwrite/fputc. Just the return value, and setting the `ferror` flag. Return value is at least a clear failure indication to app - just a bit pessimistic about where it failed.

As for IAR, um what's it doing? Can you check with some different numbers? I'd guess it's doing "successful+1" or maybe it's doing "requested-successful"? Either way, apps are in danger of false success reports (if error on last byte or first byte, respectively).

Checked the IAR docs/reference implementation, and we're doing the right thing. Doubt we could do anything to influence its behaviour, if it's calling us 1 byte at a time. Not sure why it's doing that. It's even going 1 step further and making that "0" call after every byte. According to comments in their write.c, (NULL,0) is a flush request - we should be presumably be calling sync on that. Although if it's going to do it on every byte, maybe not?

I guess they're attempting to couple fflush/nonbuffered and fsync(), but there's supposed to be a difference- fflush() flushes C library to system, and fsync() flushes system to disk. The first isn't supposed to imply the second.

Is the weird 1-at-a-time IAR behaviour maybe something to do with what mbed_set_unbuffered_stream does? I was assuming that was a bug workaround. What if you call setbuf(file, NULL) instead?

Reply:
IAR with `std::setbuf(file, NULL)` instead `mbed_set_unbuffered_stream(file)` behaves more reasonably
[log2.txt](https://github.com/ARMmbed/mbed-os/files/1677846/log2.txt)

Reply:
Weird. I actually have no idea what that IAR-specific code is supposed to achieve. It was introduced here: https://github.com/ARMmbed/mbed-os/commit/9909e7c

The `setvbuf` call there should be equivalent to `setbuf(NULL)`, but apparently isn't. Not sure what the rest of the commit affecting the behaviour of `Stream` is.

Reply:
Turns out some explanation is on GitHub, not the commit message: https://github.com/ARMmbed/mbed-os/pull/770

Is the bug now fixed, so that can be removed? What version of IAR are you testing with?